src/object/cli_obj.c

```diff
@@ -5277,36 +5308,69 @@ obj_update_shards_get(struct dc_object *obj, daos_obj_fetch_t *args, unsigned in
 {
 	uint8_t		*tgt_bitmap;
 	uint32_t	failure_cnt = 0;
+	int		grp_idx;
+	uint32_t	grp_start;
+	uint32_t	shard_nr = 0;
 	int		i;
-	int		rc;
+	int		rc = 0;
 
-	rc = obj_dkey2grpmemb(obj, obj_auxi->dkey_hash, map_ver, shard, shard_cnt);
-	if (rc || !obj_is_ec(obj))
-		return rc;
+	if (!obj_is_ec(obj))
+		return obj_dkey2grpmemb(obj, obj_auxi->dkey_hash, map_ver, shard, shard_cnt);
 
-	tgt_bitmap = obj_auxi->reasb_req.tgt_bitmap;
-	for (i = 0; i < *shard_cnt; i++) {
-		int shard_idx;
+	grp_idx = obj_dkey2grpidx(obj, obj_auxi->dkey_hash, map_ver);
+	if (grp_idx < 0)
+		return grp_idx;
 
-		if (isclr(tgt_bitmap, i))
-			continue;
+	grp_start = grp_idx * obj_get_grp_size(obj);
+	tgt_bitmap = obj_auxi->reasb_req.tgt_bitmap;
+	D_RWLOCK_RDLOCK(&obj->cob_lock);
+	for (i = 0; i < obj_get_grp_size(obj); i++) {
+		struct dc_obj_shard	*obj_shard;
+		unsigned int		shard_id;
+		int			shard_idx;
 
-		shard_idx = *shard + i;
+		shard_idx = grp_start + i;
 		D_ASSERTF(shard_idx < obj->cob_shards_nr, "%d >= %u\n",
 			  shard_idx, obj->cob_shards_nr);
 
-		if (obj_shard_is_invalid(obj, shard_idx, DAOS_OBJ_RPC_UPDATE) ||
+		obj_shard = &obj->cob_shards->do_shards[shard_idx];
+		if (obj_shard->do_target_id == -1 || obj_shard->do_shard == -1 ||
```

Comment:
if the shard is not in tgt_bitmap, need not do the fail check right?
For example EC 4+2 obj, if partial update shard 0,4,5, then need not fail the update if shard 1/2/3 failed?

Reply:
hmm, I think we do not want to update the object if it exceed the maxim failure, in case damage the parity, which will bring trouble if we need "recover" data.

Reply:
as what we discussed by skype, here with a problem that for OSA case the layout may insert some shard with -1 do_target_id (or do_shard?) that may cause fail an update which need not fail, please fix it at following patch, thx